src/MockClassifier.Api/Controllers/InstitutionController.cs

```diff
@@ -28,20 +28,25 @@ public InstitutionController(IDmrService dmrService, ITokenService tokenService,
         [HttpPost]
         public AcceptedResult Post([FromBody] MessagesInput messages)
         {
-            foreach( var message in messages.Messages)
+            if (messages == null)
+            {
+                throw new ArgumentNullException(nameof(messages));
+            }
+
+            foreach (var message in messages.Messages)
             {
                 List<string> ministries = _naturalLanguageService.Classify(message).ToList();
                 ministries = ministries.Concat(_tokenService.Classify(message).ToList()).ToList();
 
                 foreach (var ministry in ministries)
                 {
-                     _dmrService.RecordRequest(GetDmrRequest(message, ministry, messages.CallbackUri));
+                    _dmrService.RecordRequest(GetDmrRequest(message, ministry, messages.CallbackUri));
                 }
             }
             return Accepted();
         }
 
-        private static DmrRequest GetDmrRequest(string message, string ministry, string callbackUri)
+        private static DmrRequest GetDmrRequest(string message, string ministry, Uri callbackUri)
```

Comment:
Would be nice to move this logic into another class, perhaps the DmrRequest class? would be nicer design if the controller only contained API route methods

Reply:
See other responses. Refactoring to make the code better reaches beyond the scope of this PR and I don't want to make changes like this without fully evaluating the implications.